### PR TITLE
Test for Get Api Updated

### LINKED ISSUE #82: Update Tests of GET API
GET API Tests are not updated for the pagination feature, 
All the tests in controllers, managers, services shall be updated as per the latest version of GET API.

### LINKED ISSUE #107: Bug in pagination helper code
In PAgination Helper Code :-

- There are some unnecessary variable used
- Logic for total pages is also wrong
`if (numDocs &gt;= limit) { totalPages = Math.floor(numDocs / limit) + (numDocs % limit); }`
- It was not filtering Opportunities On the Basis of Opportunity-Type.

This issue has been resolve in PR of GET API TEST [#97](https://github.com/Girl-Code-It/Opportunity-Calendar-Backend/pull/97/files)

### PR DESCRIPTION
**Pull request by Sukriti-sood**
Fixes #82 
Fixes #107 
@Manvityagi 
I have added test but  , facing a problem , I am not getting idea how to stub chained mongoose Operation , I tried but only One is working (skip or limit ) not both.

### CODE REVIEW

**Comment by Manvityagi** [comment]
> Fixes #82
> 
> @Manvityagi
> I have added test but , facing a problem , I am not getting idea how to stub chained mongoose Operation , I tried but only One is working (skip or limit ) not both.

@Sukriti-sood 
Try something like this, it should work
```  
 try {
        result.results = await this.opportunity.find(
          {
            /* Everything*/
          },
          {
            /* No constraints */
          },
          {
            skip: skip,
            limit: limit,
          }
        );
   ```

**Comment by Manvityagi** [comment]
Also, pull latest code before making next commit @Sukriti-sood

**Comment by Sukriti-sood** [comment]
@Manvityagi @abdus  Added Test , Please Review.

**Comment by Sukriti-sood** [comment]
@Manvityagi  Please Review PR

**Comment by ay2306** [comment]
You are writing unit tests, you should not change the source code. Please correct it.  And your Travis Build is failing, look into it as well.

**Comment by Sukriti-sood** [comment]
> You are writing unit tests, you should not change the source code. Please correct it. And your Travis Build is failing, look into it as well.

@ay2306  Actually  Logic of Pagination Helper Was Wrong I just Corrected it.

**Comment by ay2306** [comment]
Please raise an issue, explain why is it wrong? Make PR to that. Cannot review code without context.

**Comment by Sukriti-sood** [comment]
> Please raise an issue, explain why is it wrong? Make PR to that. Cannot review code without context.

@ay2306  I have fixed the error, Can I explain it here , rather than making new issue which is not needed as changes already made I guess .


In the pagination helper variable like skip , start index , begin  was not needed  and
``` if (numDocs &gt;= limit) {    totalPages = Math.floor(numDocs / limit) + (numDocs % limit);  }```
This logic for total pages was also wrong.



Is making an Issue Needed ? I will make If its that necessary

**Comment by ay2306** [comment]
Yes, create a new issue, issues should be tracked properly, link it in this PR, and indent properly so that the diff checker shows the changes only. You have added extra characters in many places and it is difficult to see what changes have you made. Please revert all changes shown in diff which you haven't actually changed.

**Comment by Sukriti-sood** [comment]
@ay2306 Why are u asking for revert, all the changes mentioned in this PR are only made by me. and that extra characters are due to prettifying code .

**Comment by ay2306** [comment]
Yeah, dont beautify. Have to read complete file to find out changes.

**Comment by Sukriti-sood** [comment]
> Yeah, dont beautify. Have to read complete file to find out changes.

Are you Mentor Here ? As Mentor/PA of This Project believes in Clean Code as much I know.

**Comment by ay2306** [comment]
Dude, I was asked to review this PR. The issue is not with beautify, the issue is that since you beautified the whole file, it is difficult to see where are the changes. Beautify when we are sure that nothing new is added but I cannot say okay to a file without reading what is changed and it is difficult to do so because the whole file has been beautified. Imagine you are reviewing your PR and you see most of the diff say a file is changed because a tab has been added, but there are some changes as well so now you have to go line by line matching if this has been changed or not.

**Comment by Sukriti-sood** [comment]
@ay2306  I have already Opened  Issue #107 for Pagination Helper

And For File Changed , There are Only  7 files Changed and Only 2 files are extra having with  no code change ,

Still I will remove those 2 files in upcoming commit .

Anything Else needed.

**Comment by ay2306** [comment]
In `pagination.js`, remove extra space so they are not treated as changed in diff

**Comment by Sukriti-sood** [comment]
@ay2306  I have removed extra File , Please Review it

**Comment by ay2306** [comment]
You have added extra space, it is making `helper/pagination.js` like that
image
image
Remove beautification and just keep your changes there.

**Comment by Sukriti-sood** [comment]
@ay2306 Sir , where is extra space , I am not getting  your words, Please explain me.

**Comment by ay2306** [comment]
For example in `helper/pagination.js`
Earlier there was this line
image
Now that line has become this
image
Like this, there are several lines that are not changed at all but are shown as changed due to added indentation. This is not required and makes reviewing difficult because I have to compare each line one by one to see if this is something you added or this is only a tab. This defeats the whole purpose of having a diff.

Even if we review it, there will be many cases when we may need to come back to this PR in the future to see what changes were made. We need the changes to be reviewed by any developer as quickly as possible.

**Comment by Sukriti-sood** [comment]
@ay2306 Sir Please have a look now, Is this fine or still any changes needed..?

File: helpers/pagination.js [hunk 1]
```diff
@@ -49,86 +45,68 @@ export default async function pagination(queryObject, opportunity) {
...
-
-  // If any of the paramter becomes neagtive or Zero
   if (page <= 0 || limit <= 0) {
     result.results = 'The Parameters cannot be Negative or Zero';
     return result;
   }
 
-  // If user enters a Page number greater than the number of Documents present in the Database
-  if (startIndex >= numDocs) {
-    result.results = 'The maximum page value should be ' + numDocs;
```
Comment (ay2306) [inline helpers/pagination.js:71]: This is an important check, I don't think we should remove it.

File: helpers/pagination.js [hunk 2]
```diff
@@ -49,86 +45,68 @@ export default async function pagination(queryObject, opportunity) {
...
-
   // If the page entered exceeds the Max allowed value(totalPages)
   if (page > totalPages) {
     let maxAllowedPages = totalPages;
     result.results = 'The maximum allowed pages are ' + maxAllowedPages;
     return result;
   }
-
-  // StartIndex should be atleast 0 and for the min val of Page should be 1
-  if (startIndex < 0) {
```
Comment (ay2306) [inline helpers/pagination.js:83]: Why did you remove these checks?

File: helpers/pagination.js [hunk 3]
```diff
@@ -49,86 +45,68 @@ export default async function pagination(queryObject, opportunity) {
...
     }
   }
 
-  let query = {};
-  if (queryObject.type) query['opportunityType'] = queryObject.type;
-
   // Once a Valid Query(The one which is inside the Range) is entered, we render the Results
   try {
-    result.results = await opportunity.find(
-      query,
```
Comment (ay2306) [inline helpers/pagination.js]: I think we should leave it as it is, in case we require to add more to the query. Plus are you sure that `opportunityType : null` returns a list of every jobType or only for `null` ?

File: tests/fakedata.js [hunk 4]
```diff
@@ -22,4 +22,6 @@ let stubValue = {
   updatedAt: faker.date.past(),
 };
 
-export default stubValue;
+let fakedataArray = [stubValue, stubValue, stubValue, stubValue];
```
Comment (ay2306) [inline tests/fakedata.js:25]: Won't this generate an array of 4 same stubValues since stubValue is generated only once?  Please note stubValue is a variable and not a method, so here it will be same

File: tests/managers/opportunity-test.js [hunk 5]
```diff
@@ -87,6 +90,46 @@ describe('OpportunityManager', function () {
...
+        type: stubValue.opportunityType,
+        limit: -9,
+        page: -7,
+      });
+
+      expect(result.results).to.equal(
+        'The Parameters cannot be Negative or Zero'
+      );
+    });
   });
```
Comment (ay2306) [inline tests/managers/opportunity-test.js:134]: Please add tests for all situations where it will throw an error as well, like non-positive limit, non-positive page, cover more edge cases. These are only positive tests.

File: tests/services/opportunity-test.js [hunk 6]
```diff
@@ -53,11 +53,11 @@ describe('OpportunityService', function () {
...
       const stub = sinon
         .stub(opportunityManager, 'getOpportunities')
-        .returns(stubValue);
+        .returns({results:[stubValue]});
       const opportunityService = new OpportunityService(opportunityManager);
-      const opportunity = await opportunityService.getOpportunities({
+      const opportunity = (await opportunityService.getOpportunities({
         type: stubValue.opportunityType,
-      });
+      })).results[0];
```
Comment (ay2306) [inline tests/services/opportunity-test.js:60]: Add more assertions like expected length, if all entries are correct or not.

**Review by ay2306 — COMMENTED** [review]
Please look at the code review comments.

File: helpers/pagination.js (hunk 3 above)
Comment (Sukriti-sood) [reply helpers/pagination.js]: @ay2306  This has been fixed in my latest commit and I tested it by running swagger API

File: tests/fakedata.js (hunk 4 above)
Comment (Sukriti-sood) [reply tests/fakedata.js:25]: @ay2306  This is generated to test the cases like if limit is given so whats the output , and related to page query example :-
if page is 2 and limit is 2 resultant array will be of length 2.

File: tests/managers/opportunity-test.js (hunk 5 above)
Comment (Sukriti-sood) [reply tests/managers/opportunity-test.js:134]: Ook I will do.

File: helpers/pagination.js (hunk 2 above)
Comment (Sukriti-sood) [reply helpers/pagination.js:83]: Purpose this has been updated in my code.

File: tests/fakedata.js (hunk 4 above)
Comment (ay2306) [reply tests/fakedata.js:25]: What if limit is more than 4? You should add test for that as well. And wont it be better if we convert it to a function so we can call it whenever we want 6 or 10 times or even more.

Comment (Sukriti-sood) [reply tests/fakedata.js:25]: Ya it will be nyc , @ay2306 Currently I am facing issue for stubbing both mongoose operation find, skip and limit , I need help in this , .
